cmd/k8s-operator/operator.go

```diff
@@ -268,25 +270,45 @@ func childResourceLabels(parent *corev1.Service) map[string]string {
 	}
 }
 
-func (a *ServiceReconciler) Reconcile(ctx context.Context, req reconcile.Request) (_ reconcile.Result, err error) {
+func (a *ServiceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	logger := a.logger.With("service-ns", req.Namespace, "service-name", req.Name)
 	logger.Debugf("starting reconcile")
 	defer logger.Debugf("reconcile finished")
 
 	svc := new(corev1.Service)
-	err = a.Get(ctx, req.NamespacedName, svc)
-	if apierrors.IsNotFound(err) {
+	if err := a.Get(ctx, req.NamespacedName, svc); apierrors.IsNotFound(err) {
 		// Request object not found, could have been deleted after reconcile request.
 		logger.Debugf("service not found, assuming it was deleted")
 		return reconcile.Result{}, nil
 	} else if err != nil {
 		return reconcile.Result{}, fmt.Errorf("failed to get svc: %w", err)
 	}
+
+	// Add finalizer first (if not present), to avoid racing between init and deletion.
+	// If a Service resource is created and deleted (relatively quickly), before the finalizer is added,
+	// there is a risk that the deletion routine never gets called.
+	if !controllerutil.ContainsFinalizer(svc, FinalizerName) {
+		// This log line is printed exactly once during initial provisioning,
+		// because once the finalizer is in place this block gets skipped. So,
+		// this is a nice place to tell the operator that the high level,
+		// multi-reconcile operation is underway.
+		logger.Infof("exposing service over tailscale")
+		controllerutil.AddFinalizer(svc, FinalizerName)
+		if err := a.Update(ctx, svc); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to add finalizer: %w", err)
+		}
+
+		// If we just added the finalizer, we need to requeue the service to
```

Comment:
Incomplete sentence?

Reply:
Yeah, leftover before I moved the details on the outer block, will fix

Reply:
ok to close? :)